### PR TITLE
ICU-21525 UnicodeSet.hasString(), UnicodeSetIterator.skipToStrings() & C API

### PR DESCRIPTION
**Pull request by markusicu**
Also, UnicodeSetIterator is not intended for public subclassing, but there was one subclass in a unit test (in each of Java & C++). I rewrote that to compute a modified set rather than fiddle with the iterator, and removed protected @internal access in Java. In C++ we should discuss whether that's ok: ICU-21663 "UnicodeSetIterator: make protected members private"

This PR also addresses ICU-6553 "uset_charAt() / uset_size() inconsistent" via documentation and adding uset_getRangeCount().

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21525
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4j/main/classes/core/src/com/ibm/icu/number/NumberFormatter.java [hunk 1]
```diff
@@ -487,9 +487,8 @@
      * <li>AUTO: 0.90, 1.00, 1.10
      * <li>HIDE_IF_WHOLE: 0.90, 1, 1.10
      * </ul>
-     * 
+     *
```
Comment (srl295) [inline icu4j/main/classes/core/src/com/ibm/icu/number/NumberFormatter.java:490]: unrelated whitespace change

**Review by srl295 — APPROVED** [review]
LG so far

File: icu4j/main/classes/core/src/com/ibm/icu/number/NumberFormatter.java (hunk 1 above)
Comment (markusicu) [reply icu4j/main/classes/core/src/com/ibm/icu/number/NumberFormatter.java:490]: Yes, I did a drive-by removal of the last `@provisional` in ICU4J, and Eclipse did its auto-cleanup thing.

**Comment by markusicu** [comment]
The API additions were approved today, with “skipToStrings()” as in this PR, instead of my originally proposed “skipCodePoints()”. In other words, the PR already matches the approved API.

Thanks Steven for approving this PR already!

I intend to merge this near the end of today unless someone else objects.
